## Data Science Assignment 1 - Python
*By Shahequa Modabbera*

## 1. Write a Python program to reverse a string without using any built-in string reversal functions.

We can reverse a string without using any built-in string reversal functions by simply iterating through the characters of the string and constructing a new string in reverse order.

In [1]:
def reverse_string(input_str):
    reversed_str = ""

    for char in input_str:
        # Add each character to the beginning of the reversed string
        reversed_str = char + reversed_str

    return reversed_str

# Test the function
input_str = "Hello, World!"
reversed_str = reverse_string(input_str)
print("Original string:", input_str)
print("Reversed string:", reversed_str)

Original string: Hello, World!
Reversed string: !dlroW ,olleH


Explanation:
- We initialize an empty string `reversed_str` to store the reversed string.
- We then loop through each character in the original string `input_str`.
- In each iteration, we add the current character to the beginning of the `reversed_str`.
- After the loop, the `reversed_str` will contain the characters of `input_str` in reverse order.

The output shows that the original string "Hello, World!" is reversed to "!dlroW ,olleH".

## 2. Implement a function to check if a given string is a palindrome.

A palindrome is a word, phrase, number, or other sequence of characters that reads the same forward and backward. To check if a given string is a palindrome, we can compare the characters from the beginning and the end of the string, moving towards the middle.

In [2]:
def is_palindrome(input_str):
    # Remove any non-alphanumeric characters and convert to lowercase
    input_str = ''.join(char.lower() for char in input_str if char.isalnum())
    
    # Compare characters from both ends of the string
    left, right = 0, len(input_str) - 1
    while left < right:
        if input_str[left] != input_str[right]:
            return False
        left += 1
        right -= 1
    
    return True

user_input = input("Enter the text: ")
if is_palindrome(user_input):
    print("It's a palindrome!")
else:
    print("It's not a palindrome.")

Enter the text:  mam


It's a palindrome!


Explanation:
- We first remove any non-alphanumeric characters from the input string using the `isalnum()` method and convert all characters to lowercase. This is done to handle palindromes with different cases and ignore non-alphanumeric characters.
- Then, we initialize two pointers, `left` and `right`, pointing to the beginning and end of the string, respectively.
- We compare the characters at positions `left` and `right`. If they are not equal, the string is not a palindrome, and we return `False`.
- We continue the comparison until `left` is greater than or equal to `right`. If all characters match, we return `True` as the string is a palindrome.

## 3. Write a program to find the largest element in a given list.

To find the largest element in a given list, we can use the `max()` function in Python. The `max()` function returns the largest element from an iterable (e.g., list, tuple, etc.).

Explanation:
- The `find_largest_element()` function takes a list `input_list` as input.
- It uses the `max()` function to find the largest element in the list.
- The largest element is then returned as the result.

In [3]:
def find_largest_element(input_list):
    largest_element = max(input_list)
    return largest_element

# Test the function
input_list = [12, 45, 23, 67, 89, 34]
largest_element = find_largest_element(input_list)
print("The largest element in the list is:", largest_element)

The largest element in the list is: 89


Explanation:
- The `find_largest_element()` function takes a list `input_list` as input.
- It uses the `max()` function to find the largest element in the list.
- The largest element is then returned as the result.

## 4. Implement a function to count the occurrence of each element in a list.

We can implement a function that counts the occurrence of each element in a list using a dictionary to store the counts. The elements of the list will be the keys of the dictionary, and the corresponding values will be the counts.

In [4]:
def count_occurrences(input_list):
    # Create an empty dictionary to store the counts
    counts = {}

    # Iterate through the elements of the list
    for element in input_list:
        # Check if the element is already in the dictionary
        if element in counts:
            # If yes, increment the count by 1
            counts[element] += 1
        else:
            # If not, add the element to the dictionary with count 1
            counts[element] = 1

    return counts

# Test the function
input_list = [1, 2, 3, 2, 4, 1, 5, 2]
occurrences = count_occurrences(input_list)
print("Occurrences of each element:", occurrences)

Occurrences of each element: {1: 2, 2: 3, 3: 1, 4: 1, 5: 1}


Explanation:
- The `count_occurrences()` function takes a list `input_list` as input.
- It initializes an empty dictionary `counts` to store the counts of each element.
- The function then iterates through each element of the list.
- For each element, it checks if the element is already in the dictionary.
- If the element is present, it increments the count of that element by 1.
- If the element is not present, it adds the element to the dictionary with a count of 1.
- After processing all elements, the `counts` dictionary will contain the occurrence count of each element, and we return it as the result.

## 5. Write a Python program to find the second largest number in a list.

We can use the `sorted()` function in Python to find the second largest number in a list easily.

In [5]:
def find_second_largest(input_list):
    if len(input_list) < 2:
        return "List should have at least two elements."

    sorted_list = sorted(input_list, reverse=True)
    return sorted_list[1]

# Test the function
input_list = [12, 45, 23, 67, 89, 34]
second_largest = find_second_largest(input_list)
print("The second largest number in the list is:", second_largest)

The second largest number in the list is: 67


Explanation:
- The `find_second_largest()` function takes a list `input_list` as input.
- We check if the list has at least two elements. If not, we return an error message.
- We use the `sorted()` function to sort the elements of the list in descending order (`reverse=True`).
- The second largest number will be at index 1 in the sorted list (`sorted_list[1]`).
- We return the second largest number as the result.

## 6. Implement a function to remove duplicate elements from a list.

We can implement a function to remove duplicate elements from a list by converting the list to a set, which automatically removes duplicates, and then converting it back to a list.

In [6]:
def remove_duplicates(input_list):
    # Convert the list to a set to remove duplicates
    unique_elements = set(input_list)

    # Convert the set back to a list
    result_list = list(unique_elements)

    return result_list

# Test the function
input_list = [1, 2, 2, 3, 3, 4, 4, 5]
result_list = remove_duplicates(input_list)
print("List with duplicates removed:", result_list)

List with duplicates removed: [1, 2, 3, 4, 5]


Explanation:
- The `remove_duplicates()` function takes a list `input_list` as input.
- We convert the list to a set `unique_elements`, which automatically removes duplicates since sets do not allow duplicate elements.
- We then convert the set `unique_elements` back to a list `result_list`.
- The `result_list` will contain only the unique elements from the original list, and we return it as the result.

## 7. Write a program to calculate the factorial of a given number.

To calculate the factorial of a given number, we can use a simple loop to multiply all the numbers from 1 to the given number.

In [7]:
def factorial(n):
    if n < 0:
        return "Factorial is not defined for negative numbers."
    elif n == 0 or n == 1:
        return 1
    else:
        result = 1
        for i in range(1, n + 1):
            result *= i
        return result

# Test the function
number = 5
print(f"The factorial of {number} is: {factorial(number)}")

The factorial of 5 is: 120


Explanation:
- The `factorial()` function takes an integer `n` as input.
- It first checks if `n` is negative, and if so, it returns an error message.
- If `n` is 0 or 1, the factorial is 1, so we directly return 1 for these cases.
- For all other positive values of `n`, we initialize a variable `result` to 1.
- We then use a loop to multiply all the numbers from 1 to `n` to calculate the factorial.
- After the loop, the `result` will contain the factorial of the given number, and we return it as the result.

In the example, the factorial of 5 is calculated as 5 * 4 * 3 * 2 * 1 = 120, so the output is `The factorial of 5 is: 120`.

## 8. Implement a function to check if a given number is prime.

To check if a given number is prime, we need to verify if it is only divisible by 1 and itself. We can achieve this by iterating from 2 to the square root of the number and checking if there is any divisor.

In [8]:
import math

def is_prime(number):
    if number < 2:
        return False

    for divisor in range(2, int(math.sqrt(number)) + 1):
        if number % divisor == 0:
            return False

    return True

# Test the function
number = 17
if is_prime(number):
    print(f"{number} is a prime number.")
else:
    print(f"{number} is not a prime number.")

17 is a prime number.


Explanation:
- The `is_prime()` function takes an integer `number` as input.
- It first checks if the number is less than 2. If it is, the function returns `False`, as numbers less than 2 are not prime.
- Next, it iterates from 2 to the square root of the number using a loop.
- For each divisor in this range, it checks if the number is divisible by the divisor without a remainder.
- If any divisor is found, the function returns `False`, indicating that the number is not prime.
- If no divisor is found, the function returns `True`, indicating that the number is prime.

In the example, the number 17 is a prime number because it is only divisible by 1 and 17 itself, so the output is `17 is a prime number.`.

## 9. Write a Python program to sort a list of integers in ascending order.

We can use the built-in `sorted()` function or the `sort()` method to sort a list of integers in ascending order.

1. Using the `sorted()` function:

In [9]:
def sort_list_ascending(input_list):
    sorted_list = sorted(input_list)
    return sorted_list

# Test the function
numbers = [5, 2, 8, 1, 3]
sorted_numbers = sort_list_ascending(numbers)
print("Sorted list:", sorted_numbers)

Sorted list: [1, 2, 3, 5, 8]


2. Using the `sort()` method (which sorts the list in-place):

In [10]:
def sort_list_ascending(input_list):
    input_list.sort()
    return input_list

# Test the function
numbers = [5, 2, 8, 1, 3]
sorted_numbers = sort_list_ascending(numbers)
print("Sorted list:", sorted_numbers)

Sorted list: [1, 2, 3, 5, 8]


Both methods will give us a sorted list of integers in ascending order. We can choose either of them based on whether we want to create a new sorted list (`sorted()` function) or sort the original list in place (`sort()` method).

## 10. Implement a function to find the sum of all numbers in a list.

We can implement a function to find the sum of all numbers in a list by using a loop to iterate through the list and adding each element to a running total. Here's how you can do it:

In [11]:
def find_sum(numbers):
    total = 0
    for num in numbers:
        total += num
    return total

# Test the function
numbers = [1, 2, 3, 4, 5]
result = find_sum(numbers)
print("Sum of numbers:", result)

Sum of numbers: 15


Explanation:
- The `find_sum()` function takes a list `numbers` as input.
- It initializes a variable `total` to 0, which will store the sum of all numbers.
- It then uses a loop to iterate through each element `num` in the `numbers` list.
- Inside the loop, it adds the current element `num` to the `total` variable.
- After the loop, the function returns the final value of `total`, which represents the sum of all numbers in the list.

## 11. Write a program to find the common elements between two lists.

We can find the common elements between two lists by using a set data structure in Python. Sets automatically remove duplicate elements, and they provide fast membership testing, making it efficient to find common elements.

In [12]:
def find_common_elements(list1, list2):
    # Convert both lists to sets
    set1 = set(list1)
    set2 = set(list2)
    
    # Find the intersection of the two sets (common elements)
    common_elements = set1.intersection(set2)
    
    return list(common_elements)

# Test the function
list1 = [1, 2, 3, 4, 5]
list2 = [3, 4, 5, 6, 7]
result = find_common_elements(list1, list2)
print("Common elements:", result)

Common elements: [3, 4, 5]


Explanation:
- The `find_common_elements()` function takes two lists, `list1` and `list2`, as input.
- It first converts both lists into sets using the `set()` function, which automatically removes duplicate elements.
- Next, it finds the intersection of the two sets using the `intersection()` method, which gives us the common elements between the two sets.
- Finally, the function converts the resulting set back into a list and returns it as the output.

## 12. Implement a function to check if a given string is an anagram of another string.

An anagram is a word or phrase formed by rearranging the letters of another word or phrase, typically using all the original letters exactly once. To check if a given string is an anagram of another string, we can compare if both strings have the same characters with the same frequency.

In [13]:
def is_anagram(str1, str2):
    # Convert both strings to lowercase to handle case-insensitivity
    str1 = str1.lower()
    str2 = str2.lower()
    
    # Check if the two strings have the same length
    if len(str1) != len(str2):
        return False

    # Convert both strings to dictionaries containing character frequencies
    char_freq1 = {}
    char_freq2 = {}

    for char in str1:
        char_freq1[char] = char_freq1.get(char, 0) + 1

    for char in str2:
        char_freq2[char] = char_freq2.get(char, 0) + 1

    # Compare the two dictionaries
    return char_freq1 == char_freq2

# Test the function
string1 = "listen"
string2 = "silent"
if is_anagram(string1, string2):
    print("The strings are anagrams.")
else:
    print("The strings are not anagrams.")

The strings are anagrams.


Explanation:
- The `is_anagram()` function takes two strings, `str1` and `str2`, as input.
- It first converts both strings to lowercase using the `lower()` method to handle case-insensitivity. For anagrams, the case of characters doesn't matter.
- It then checks if the lengths of the two strings are the same. If not, they cannot be anagrams, so the function returns `False`.
- Next, it converts both strings to dictionaries containing the frequency of each character in the string. The keys in the dictionary are characters, and the values are their frequencies.
- Finally, it compares the two dictionaries. If they are equal, it means that both strings have the same characters with the same frequency, and hence they are anagrams. The function returns `True` in this case.

## 13. Write a Python program to generate all permutations of a given string.

We can use the `itertools.permutations` function to generate all permutations of a given string.

In [14]:
from itertools import permutations

def generate_permutations(input_string):
    # Use itertools.permutations to generate all permutations
    perms = permutations(input_string)
    
    # Convert each permutation tuple to a string and store them in a list
    result = [''.join(perm) for perm in perms]
    
    return result

# Test the function
input_string = "abc"
permutations_list = generate_permutations(input_string)
print("Permutations of", input_string, ":", permutations_list)

Permutations of abc : ['abc', 'acb', 'bac', 'bca', 'cab', 'cba']


Explanation:
- The `generate_permutations()` function takes a string `input_string` as input.
- It uses the `itertools.permutations` function to generate all possible permutations of the characters in the `input_string`.
- The `permutations()` function returns an iterator of tuples, where each tuple represents a permutation of characters.
- The function then converts each tuple to a string using the `''.join(perm)` method and stores the strings in the `result` list.
- Finally, it returns the list of all permutations as the output.

## 14. Implement a function to calculate the Fibonacci sequence up to a given number of terms.

The Fibonacci sequence is a series of numbers in which each number is the sum of the two preceding ones, usually starting with 0 and 1. 

In [15]:
def fibonacci_sequence(n):
    fib_sequence = [0, 1]  # Initialize the Fibonacci sequence with the first two terms
    while len(fib_sequence) < n:
        next_term = fib_sequence[-1] + fib_sequence[-2]  # Calculate the next term
        fib_sequence.append(next_term)
    return fib_sequence

# Test the function
n = 10
fib_seq = fibonacci_sequence(n)
print("Fibonacci sequence with", n, "terms:", fib_seq)

Fibonacci sequence with 10 terms: [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]


Explanation:
- The `fibonacci_sequence()` function takes an integer `n` as input, which represents the number of terms we want in the Fibonacci sequence.
- It initializes the `fib_sequence` list with the first two terms `[0, 1]`, which are the starting points of the Fibonacci sequence.
- It then enters a `while` loop, which continues until the length of `fib_sequence` is equal to `n`.
- Inside the loop, it calculates the next term in the Fibonacci sequence by summing the last two elements of the `fib_sequence`.
- The function appends the next term to the `fib_sequence` list.
- The loop continues until the required number of terms `n` is reached.
- Finally, the function returns the `fib_sequence` list as the output.

## 15. Write a program to find the median of a list of numbers.

To find the median of a list of numbers, we can follow these steps:

1. Sort the list in ascending order (if it's not already sorted).
2. If the list has an odd number of elements, the median is the middle element.
3. If the list has an even number of elements, the median is the average of the two middle elements.

In [16]:
def find_median(numbers):
    # Sort the list in ascending order
    sorted_numbers = sorted(numbers)
    n = len(sorted_numbers)
    
    # Check if the number of elements is odd or even
    if n % 2 == 1:
        # If odd, return the middle element
        median = sorted_numbers[n//2]
    else:
        # If even, return the average of the middle elements
        middle_right = sorted_numbers[n//2]
        middle_left = sorted_numbers[(n//2) - 1]
        median = (middle_left + middle_right) / 2
        
    return median

# test the function
numbers = [3,1,5,3,2,4]
median_value = find_median(numbers)
print("Median:", median_value)

Median: 3.0


Explanation:
- The `find_median()` function takes a list of numbers as input.
- It first sorts the list using the `sorted()` function to ensure that the elements are in ascending order.
- It then calculates the length of the sorted list, `n`.
- If `n` is odd, the function returns the middle element of the sorted list as the median.
- If `n` is even, the function calculates the average of the two middle elements and returns it as the median.

## 16. Implement a function to check if a given list is sorted in non-decreasing order.

We can implement a function to check if a given list is sorted in non-decreasing order by iterating through the list and comparing each element with the next one.

In [17]:
def is_sorted_non_decreasing(lst):
    n = len(lst)
    for i in range(n-1):
        if lst[i] > lst[i+1]:
            return False
    return True
    
# Test the function
list1 = [1,2,3,4,5]
list2 = [3,4,7,4,2,1]
        
print("List1 is sorted:", is_sorted_non_decreasing(list1))
print("List2 is sorted:", is_sorted_non_decreasing(list2))

List1 is sorted: True
List2 is sorted: False


Explanation:
- The `is_sorted_non_decreasing()` function takes a list `lst` as input.
- It calculates the length of the list `n`.
- Then, it iterates through the list using a `for` loop from index 0 to `n-2` (i.e., the second-to-last element).
- Inside the loop, it compares each element `lst[i]` with the next element `lst[i + 1]`.
- If at any point `lst[i]` is greater than `lst[i + 1]`, it means the list is not sorted in non-decreasing order, and the function returns `False`.
- If the loop completes without finding any out-of-order elements, it means the list is sorted in non-decreasing order, and the function returns `True`.

## 17. Write a Python program to find the intersection of two lists.

We can find the intersection of two lists in Python using the built-in `set` data type and its `intersection()` method.

In [18]:
def find_intersection(list1, list2):
    # Convert both lists to sets
    set1 = set(list1)
    set2 = set(list2)
    
    # Find the intersection of the two sets
    intersection_set = set1.intersection(set2)
    
    # Convert the intersection set back to list
    intersection_list = list(intersection_set)
    
    return intersection_list

# test the function
list1 = [1,2,3,4,5, 7]
list2 = [3,5,1,7,8,2]

intersection_result = find_intersection(list1, list2)
print("Intersection of the two lists:", intersection_result)

Intersection of the two lists: [1, 2, 3, 5, 7]


Explanation:
- The `find_intersection()` function takes two lists `list1` and `list2` as input.
- It converts both lists to sets using the `set()` constructor, which automatically removes duplicate elements and performs set operations efficiently.
- Then, it uses the `intersection()` method on one of the sets (in this case, `set1`) and provides the other set (`set2`) as the argument to find their intersection.
- The result of the intersection operation is a set containing the common elements present in both `set1` and `set2`.
- Finally, the function converts the intersection set back to a list using the `list()` constructor and returns it as the output.

## 18. Implement a function to find the maximum subarray sum in a given list.

To find the maximum subarray sum in a given list, you can use Kadane's algorithm, which has a time complexity of O(n)

In [19]:
def max_subarray_sum(nums):
    max_sum = float('-inf') # Initialize max_sum to negative infinity
    current_sum = 0 # Initialize current_sum to 0
    
    for num in nums:
        # add the current number to the current_sum
        current_sum += num
        
        # if current_sum becomes negative, reset it to 0
        if current_sum < 0:
            current_sum = 0
            
        # Update max_sum if current_sum is greater
        if current_sum > max_sum:
            max_sum = current_sum
            
    return max_sum
            
# test the function
numbers = [2, 1, -3, 7, -6, 1, -1, 2, -3, 4, -5, 6]
print("Maximum subarray sum:", max_subarray_sum(numbers))

Maximum subarray sum: 7


Explanation:
- The `max_subarray_sum()` function takes a list `nums` as input.
- It initializes two variables: `max_sum` to negative infinity and `current_sum` to 0.
- The function then iterates through the list `nums`.
- For each element in `nums`, it adds the element to the `current_sum`.
- If `current_sum` becomes negative, it means that including that element in the subarray would not give us the maximum sum, so we reset `current_sum` to 0 to start a new subarray from the next element.
- If `current_sum` is greater than `max_sum`, it means we have found a new maximum subarray sum, so we update `max_sum`.
- After the loop completes, the function returns `max_sum`, which represents the maximum subarray sum.

## 19. Write a program to remove all vowels from a given string.

In [20]:
def remove_vowels(input_str):
    vowels = 'AEIOUaeiou'
    result_str = ""
    
    for char in input_str:
        if char not in vowels:
            result_str += char
            
    return result_str

# Test the function
input_string = 'Hello, Shahequa!'
print("Input string:", input_string)
print("String after removing vowels:", remove_vowels(input_string))

Input string: Hello, Shahequa!
String after removing vowels: Hll, Shhq!


Explanation:
- The `remove_vowels()` function takes an input string `input_str` as an argument.
- We define a string `vowels` containing all the uppercase and lowercase vowels.
- We initialize an empty string `result_str`, which will store the characters of the input string that are not vowels.
- We then loop through each character in the input string using a `for` loop.
- If the character is not present in the `vowels` string (i.e., it is not a vowel), we add it to the `result_str`.
- After processing all characters in the input string, the function returns the `result_str`, which is the input string with all vowels removed.

## 20. Implement a function to reverse the order of words in a given sentence.

To reverse the order of words in a given sentence, we can split the sentence into individual words, reverse the list of words, and then join them back into a new sentence.

In [21]:
def reverse_words(sentence):
    # Split the sentence into individual words
    words_list = sentence.split()
    
    # reverse the list of words
    reversed_list = words_list[::-1]
    
    # join the reversed words to form a new sentence
    reverse_sentence = " ".join(reversed_list)
    
    return reverse_sentence

# test the function
input_sentence = "Hello, Shahequa! You are amazing!"
print("Input Senetence:", input_sentence)
print("Reversed Sentence:", reverse_words(input_sentence))

Input Senetence: Hello, Shahequa! You are amazing!
Reversed Sentence: amazing! are You Shahequa! Hello,


Explanation:
- The `reverse_words()` function takes a sentence as input and returns the reversed sentence.
- We first split the input sentence into individual words using the `split()` method, which returns a list of words in the sentence (`words_list`).
- We then use slicing (`[::-1]`) to reverse the list of words (`reversed_list`).
- Finally, we join the reversed list of words back into a new sentence using the `join()` method and return the result (`reversed_sentence`).

## 21. Write a Python program to check if two strings are anagrams of each other.

In [22]:
def are_anagrams(str1, str2):
    # Convert both strings to lowercase to make the comparison case-sensitive
    str1 = str1.lower()
    str2 = str2.lower()
    
    # remove any spaces from both strings
    str1 = str1.replace(" ", "")
    str2 = str2.replace(" ", "")
    
    # Check if the sorted characters of both strings are the same
    return sorted(str1) == sorted(str2)

# test the function
string1 = "Listen"
string2 = "Silent"
print("String 1:", string1)
print("String 2:", string2)
print("Are they anagrams?", are_anagrams(string1, string2))

String 1: Listen
String 2: Silent
Are they anagrams? True


Explanation:
- The `are_anagrams()` function takes two strings (`str1` and `str2`) as input and returns a boolean value indicating whether they are anagrams of each other.
- We convert both strings to lowercase using the `lower()` method to make the comparison case-insensitive.
- We remove any spaces from both strings using the `replace()` method.
- We then check if the sorted characters of both strings are the same. If they are, it means the strings are anagrams, and the function returns `True`; otherwise, it returns `False`.

## 22. Implement a function to find the first non-repeating character in a string.

In [23]:
def first_non_repeating_char(input_str):
    char_count = {}
    
    # Count the occurences of each character in the string
    for char in input_str:
        char_count[char] = char_count.get(char,0) + 1
        
    # Find the first non-repeating character
    for char in input_str:
        if char_count[char] == 1:
            return char
        
    # If no non-repeating character found, return None
    return None

# test the function
input_string = "abacbcddeefft"
result = first_non_repeating_char(input_string)
if result is not None:
    print("First non-repeating character:", result)
else:
    print("No non-repeating character found.")

First non-repeating character: t


Explanation:
- The `first_non_repeating_char()` function takes a string `input_str` as input and returns the first non-repeating character in the string.
- We create a dictionary `char_count` to store the count of each character in the string.
- We loop through each character in the input string and update its count in the dictionary.
- After counting the occurrences of each character, we loop through the string again and check the count of each character in the dictionary.
- If the count of a character is equal to 1, it means it is the first non-repeating character, and we return it.
- If no non-repeating character is found, we return `None`.

## 23. Write a program to find the prime factors of a given number.

In [24]:
def is_prime(num):
    if num <= 1:
        return False
    for i in range(2, int(num ** 0.5) + 1):
        if num % i == 0:
            return False
    return True

def prime_factors(number):
    factors = []
    if number <= 1:
        return factors
    
    for i in range(2, number + 1):
        if is_prime(i) and number % i == 0:
            factors.append(i)
            
    return factors

# Test the function
num = 72
print("Prime factors of", num, ":", prime_factors(num))

Prime factors of 72 : [2, 3]


Explanation:
- The `is_prime()` function is a helper function that checks whether a given number is prime or not. It returns `True` if the number is prime and `False` otherwise.
- The `prime_factors()` function takes a number as input and returns a list of its prime factors.
- We first create an empty list `factors` to store the prime factors.
- We then iterate through numbers from 2 to the given number (inclusive).
- For each number in the range, we check if it is a prime number using the `is_prime()` function and if it is a factor of the given number (i.e., the given number is divisible by it). If both conditions are true, we add the number to the `factors` list.
- Finally, we return the list of prime factors.

## 24. Implement a function to check if a given number is a power of two.

In [25]:
def is_power_of_two(number):
    if number <= 0:
        return False
    return number & (number - 1) == 0

# Test the function
num = 16
if is_power_of_two(num):
    print(num, "is a power of two.")
else:
    print(num, "is not a power of two.")

16 is a power of two.


Explanation:
- The `is_power_of_two()` function takes a number as input and returns `True` if the number is a power of two, otherwise `False`.
- A power of two in binary representation has only one bit set to 1 (e.g., 2^0 = 1, 2^1 = 10, 2^2 = 100, 2^3 = 1000, etc.).
- If we subtract 1 from a power of two, all its bits will be set to 1 except for the most significant bit that was set to 1 in the original number (e.g., 2^3 = 1000, 2^3 - 1 = 0111).
- When we perform the bitwise AND operation between a power of two and its subtracted value, the result will be 0, indicating that the number is a power of two.

## 25. Write a Python program to merge two sorted lists into a single sorted list.

In [26]:
def merge_sorted_lists(list1, list2):
    merged_list = []
    i, j = 0, 0

    while i < len(list1) and j < len(list2):
        if list1[i] < list2[j]:
            merged_list.append(list1[i])
            i += 1
        else:
            merged_list.append(list2[j])
            j += 1

    # Append remaining elements from both lists
    merged_list.extend(list1[i:])
    merged_list.extend(list2[j:])

    return merged_list

# Sorted lists to merge
list1 = [1, 3, 5, 7, 9]
list2 = [2, 4, 6, 8, 10]

# Merge the lists
merged_result = merge_sorted_lists(list1, list2)

# Print the merged result
print("Merged Sorted List:", merged_result)

Merged Sorted List: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


Explanation:
- The `merge_sorted_lists()` function takes two sorted lists `list1` and `list2` as input and returns a new list that contains all the elements from both lists in sorted order.
- We use two pointers `i` and `j` to track the current positions in `list1` and `list2`, respectively.
- We compare the elements at the current positions `i` and `j` in both lists and append the smaller element to the `merged_list`. Then, we move the corresponding pointer one step forward.
- We continue this process until we reach the end of either `list1` or `list2`.
- Finally, we append the remaining elements from the other list to the `merged_list`.
- The `merged_list` is then returned as the output.

## 26. Implement a function to find the mode of a list of numbers.

The mode of a list of numbers is the number that appears most frequently in the list.

In [27]:
def find_mode(numbers):
    freq_dict = {}
    max_freq = 0
    modes = []
    
    # Count the occurences of each numbers in the list
    for num in numbers:
        freq_dict[num] = freq_dict.get(num, 0) + 1
        max_freq = max(max_freq, freq_dict[num])
        
    # Find all numbers with the max frequency (modes)
    for num, freq in freq_dict.items():
        if freq == max_freq:
            modes.append(num)
    
    return modes

# Test the function
numbers_list = [2,3,5,7,8,5,5,4,3,2,5,7,1]
result = find_mode(numbers_list)
print("Mode(s) of the list:", result)

Mode(s) of the list: [5]


Explanation:
- The `find_mode()` function takes a list of numbers `numbers` as input and returns a list of all the modes (numbers with the maximum frequency) in the input list.
- We create a dictionary `freq_dict` to store the count of occurrences of each number in the list.
- We also keep track of the maximum frequency (`max_freq`) found so far.
- We iterate through the list and update the count in the dictionary for each number.
- At the same time, we update the `max_freq` with the maximum count found.
- After counting the occurrences of each number, we iterate through the dictionary to find all numbers that have the maximum frequency and add them to the `modes` list.
- Finally, we return the list of modes.

## 27. Write a program to find the greatest common divisor (GCD) of two numbers.

Here's a Python program to find the greatest common divisor (GCD) of two numbers using the Euclidean algorithm:

In [28]:
def gcd(a, b):
    while b != 0:
        a, b = b, a % b
    return a

# test the function
num1 = 24
num2 = 36
result = gcd(num1, num2)
print("GCD of", num1, "and", num2, "is:", result)

GCD of 24 and 36 is: 12


Explanation:
- The `gcd()` function takes two numbers `a` and `b` as input and returns their greatest common divisor (GCD).
- The function uses the Euclidean algorithm, which repeatedly divides the larger number by the smaller number and updates the numbers until the remainder becomes 0.
- The GCD is the value of the larger number when the remainder becomes 0.

## 28. Implement a function to calculate the square root of a given number.

To calculate the square root of a given number we can use the `math.sqrt()` function from the Python `math` module.

In [29]:
import math

def square_root(number):
    if number < 0:
        raise ValueError("Square root is not defined for negative numbers")
        
    return math.sqrt(number)

# test the function
number = 25
result = square_root(number)
print("Square root of", number, "is:", result)

Square root of 25 is: 5.0


 Explanation:
- In this approach, we utilize the `math.sqrt()` function from the `math` module in Python.
- The `math.sqrt()` function takes a non-negative number as input and returns its square root as a floating-point number.
- If the input number is negative, the function raises a `ValueError` to indicate that the square root is not defined for negative numbers.

## 29. Write a Python program to check if a given string is a valid palindrome ignoring non-alphanumeric characters.

In [30]:
import re

def is_palindrome(s):
    # Remove non=alphnumeric characters and convert to lowercase
    s = re.sub(r'\W+', '', s).lower()
    
    # Check if the string is a palindrome
    return s == s[::-1]

# test the function
input_strng = "A man, a plan, a canal: Panama"
print("Is the given string a valid palindrome?", is_palindrome(input_strng))

Is the given string a valid palindrome? True


Explanation:
- The `is_valid_palindrome()` function takes a string `s` as input and returns `True` if the string is a valid palindrome after removing non-alphanumeric characters and converting to lowercase, otherwise it returns `False`.
- The `re.sub(r'\W+', '', s).lower()` part of the code uses a regular expression (`\W+`) to remove all non-alphanumeric characters from the string `s`, and then converts the resulting string to lowercase using the `lower()` method.
- After removing non-alphanumeric characters and converting to lowercase, we check if the resulting string is equal to its reverse (`s == s[::-1]`), which indicates whether the original string is a palindrome.

## 30. Implement a function to find the minimum element in a rotated sorted list.

To find the minimum element in a rotated sorted list, you can use a modified binary search algorithm.

In [31]:
def find_min_in_rotated_list(nums):
    left, right = 0, len(nums) - 1
    
    while left < right:
        mid = left + (right - left) // 2
        
        if nums[mid] > nums[right]:
            left = mid + 1
        else:
            right = mid
            
    return nums[left]

# Test the function
rotated_list = [4,5,6,7,0,1,2]
result = find_min_in_rotated_list(rotated_list)
print("Minimum element in the rotated sorted list:", result)

Minimum element in the rotated sorted list: 0


Explanation:
- The `find_min_in_rotated_list()` function takes a rotated sorted list `nums` as input and returns the minimum element in the list.
- We initialize two pointers, `left` and `right`, which represent the start and end of the current search range.
- The function uses a modified binary search algorithm to find the minimum element. In each iteration of the while loop, it calculates the mid index of the current range.
- If `nums[mid] > nums[right]`, it means the minimum element must be on the right side of the mid index, so we update `left` to `mid + 1`.
- Otherwise, the minimum element must be on the left side of the mid index, so we update `right` to `mid`.
- The loop continues until `left` and `right` point to the same element, which is the minimum element in the rotated sorted list.

## 31. Write a program to find the sum of all even numbers in a list.

In [32]:
def sum_of_even_numbers(nums):
    # Initialize a variable to hold the sum of even numbers
    total = 0
    
    # Iterate through the list
    for num in nums:
        # Check if the number is even
        if num % 2 == 0:
            # Add the even number to the total sum
            total += num
    
    return total

# test the function
numbers_list = [1,2,3,4,5,6,7,8,9,10]
result = sum_of_even_numbers(numbers_list)
print("Sum of even numbers in the list:", result)

Sum of even numbers in the list: 30


Explanation:
- The `sum_of_even_numbers()` function takes a list of numbers `nums` as input and returns the sum of all even numbers in the list.
- We initialize a variable `total` to 0 to store the sum of even numbers.
- The function iterates through the list using a `for` loop and checks each number to see if it is even (i.e., its remainder when divided by 2 is 0).
- If a number is even, we add it to the `total` sum using the `+=` operator.
- After iterating through all numbers in the list, the function returns the final sum of even numbers.

## 32. Implement a function to calculate the power of a number using recursion.

To calculate the power of a number using recursion, we can implement a recursive function that multiplies the base number with itself recursively until the desired power is reached.

In [33]:
def power(base, exponent):
    # base case: when the exponent is 0, the result is 1
    if exponent == 0:
        return 1
    # Recursive case: calculate power using recursion
    else:
        return base * power(base, exponent - 1)
    
# test the function
base_number = 2
exponent_number = 3
result = power(base_number, exponent_number)
print(f"{base_number} raised to the power {exponent_number} is:", result)

2 raised to the power 3 is: 8


Explanation:
- The `power()` function takes two arguments: `base` and `exponent`, representing the base number and the power to which it needs to be raised.
- The function uses recursion to calculate the power. It has a base case where the exponent is 0, and in that case, the function returns 1 (any number raised to the power of 0 is 1).
- In the recursive case, the function returns the product of the base number and the result of calling itself with the base number and the decremented exponent (`base * power(base, exponent - 1)`).
- The recursion continues until the exponent becomes 0, at which point the base case is triggered, and the recursion unwinds to return the final result.

## 33. Write a Python program to remove duplicates from a list while preserving the order.

In [34]:
def remove_duplicates(input_list):
    # Initialize an empty list to store unique elements while preserving order
    unique_list =[]
    
    # Iterate through the input list
    for item in input_list:
        # If the item is not already in the unique list, add it
        if item not in unique_list:
            unique_list.append(item)
            
    return unique_list

# test the function
input_list = [1,2,3,4,5,6,7,8,4,1,3,2]
result = remove_duplicates(input_list)
print("List with duplicates removed:", result)

List with duplicates removed: [1, 2, 3, 4, 5, 6, 7, 8]


Explanation:
- The `remove_duplicates()` function takes a list `input_list` as input and returns a new list containing only the unique elements while preserving their order.
- We initialize an empty list `unique_list` to store the unique elements.
- The function iterates through the input list using a `for` loop and checks each item to see if it is already present in the `unique_list`.
- If the item is not already in the `unique_list`, we add it using the `append()` method, ensuring that only the first occurrence of each element is added to the list.
- After processing all elements in the input list, the function returns the `unique_list` containing only the unique elements in their original order.

## 34. Implement a function to find the longest common prefix among a list of strings.

In [35]:
def longest_common_prefix(strs):
    # Check if the list is empty
    if not strs:
        return ""
    
    # Initialize the longest common prefix as the first string in the list
    prefix = strs[0]
    
    #Iterate through the remaining strings in the list
    for s in strs[1:]:
        #Find the common prefix between the current prefix and the current string
        i = 0
        while i < len(prefix) and i < len(s) and prefix[i] == s[i]:
            i += 1
        prefix = prefix[:i]
        
    return prefix

# Test the function
strings_list = ["flower", "flow", "flight", "float"]
result = longest_common_prefix(strings_list)
print("Longest common prefix:", result)

Longest common prefix: fl


Explanation:
- The `longest_common_prefix()` function takes a list of strings `strs` as input and returns the longest common prefix among them.
- We first check if the list is empty. If it is, there is no common prefix, so we return an empty string.
- We initialize the `prefix` variable with the first string in the list.
- The function then iterates through the remaining strings in the list using a `for` loop.
- For each string, we find the common prefix between the current `prefix` and the string by comparing characters at each position until a mismatch is found or one of the strings is exhausted.
- The common prefix is updated at each step, and after processing all strings in the list, the final value of `prefix` will be the longest common prefix.

## 35. Write a program to check if a given number is a perfect square.

To check if a given number is a perfect square, we can use the property that a perfect square is an integer that is equal to the square of another integer. 

In [36]:
def is_perfect_square(num):
    #base case: check if the number is less than 0
    if num < 0:
        return False
    
    # Use the property that a perfect square is an integer whose square root is an integer
    sqrt_num = int(num ** 0.5)
    return sqrt_num * sqrt_num == num

# Test the function
number = 25
if is_perfect_square(number):
    print(f"{number} is a perfect square")
else:
    print(f"{number} is not a perfect square")

25 is a perfect square


Explanation:
- The `is_perfect_square()` function takes an integer `num` as input and returns `True` if `num` is a perfect square, and `False` otherwise.
- We first check the base case to ensure that the number is non-negative. A negative number cannot be a perfect square since squaring a negative number gives a positive result.
- We then calculate the square root of the number using the expression `num ** 0.5`. The `int()` function is used to convert the result to an integer, truncating any decimal part.
- Finally, we check if the square of the square root (`sqrt_num * sqrt_num`) is equal to the original number. If they are equal, it means the number is a perfect square.

## 36. Implement a function to calculate the product of all elements in a list.

In [37]:
def product_of_elements(input_list):
    # Initialize the product as 1
    product = 1
    
    # Iterate through the elements in the list
    for num in input_list:
        # Multiply each element to the current product
        product *= num
        
    return product

# Test the function
numbers = [2,3,4,5]
result = product_of_elements(numbers)
print("Product of elements:", result)

Product of elements: 120


Explanation:
- The `product_of_elements()` function takes a list `input_list` as input and returns the product of all elements in the list.
- We initialize the `product` variable to 1, as multiplying by 1 does not change the value.
- The function then iterates through each element in the list using a `for` loop.
- For each element, we multiply it with the current `product` value using the `*=` operator. This updates the `product` to include the contribution of the current element.
- After processing all elements in the list, the final value of `product` will be the product of all elements in the list.

## 37. Write a Python program to reverse the order of words in a sentence while preserving the word order.

In [38]:
def reverse_sentence(sentence):
    # Split the sentence into words using whitespace as the delimiter
    words = sentence.split()
    
    # Reverse the list of words
    reversed_words = words[::-1]
    
    # Join the reversed words to form the reversed sentence
    reversed_sentence = " ".join (reversed_words)
    
    return reversed_sentence

# test the function
input_sentence = "Hello world, how are you?"
reversed_sentence = reverse_sentence(input_sentence)
print("Reversed senetence:", reversed_sentence)

Reversed senetence: you? are how world, Hello


Explanation:
- The `reverse_sentence()` function takes a sentence `sentence` as input and returns the reversed version of the sentence while preserving the word order.
- We first split the input sentence into a list of words using the `split()` method. The default delimiter is whitespace, so it will separate the words based on spaces.
- We then reverse the list of words using slicing notation `[::-1]`. This creates a reversed copy of the list without modifying the original list.
- Finally, we join the reversed list of words back into a single string using the `join()` method with whitespace as the delimiter.

## 38. Implement a function to find the missing number in a given list of consecutive numbers.

To find the missing number in a given list of consecutive numbers, we can use the formula for the sum of an arithmetic sequence and then compare it with the sum of the actual elements in the list. The difference between these sums will be the missing number.

In [39]:
def find_missing_number(nums):
    # calculate the expected sum of the consecutive nnumbers using the formula for an arithmetic sequence
    n = len(nums) + 1 # Number of elements in the original sequence (including the missing number)
    expected_sum = (n * (n+1)) // 2
    
    # calculete the actual sum of the elements in the list
    actual_sum = sum(nums)
    
    # The difference between the expected and actual sum will be the missing number
    missing_number = expected_sum - actual_sum
    
    return missing_number

# test the function
numbers = [1,2,3,4,6]
missing_number = find_missing_number(numbers)
print("Missing number:", missing_number)

Missing number: 5


Explanation:
- The `find_missing_number()` function takes a list of consecutive numbers `nums` as input and returns the missing number in the sequence.
- We calculate the expected sum of the consecutive numbers using the formula for an arithmetic sequence, where `n` is the number of elements in the original sequence (including the missing number).
- We then calculate the actual sum of the elements in the list using the `sum()` function.
- The missing number is the difference between the expected and actual sum.

## 39. Write a program to find the sum of digits of a given number.

In [40]:
def sum_of_digits(number):
    # Initialize the sum to 0
    sum_digits = 0
    
    # Convert the number to a string to iterate through its digits
    num_str = str(number)
    
    # Iterate through each digit in the string
    for digit in num_str:
        # Convert the digit back to an integer and add it to the sum
        sum_digits += int(digit)
        
    return sum_digits

#test the function
num = 12345
result = sum_of_digits(num)
print("Sum of digits:", result)

Sum of digits: 15


Explanation:
- The `sum_of_digits()` function takes a number `number` as input and returns the sum of its digits.
- We initialize the `sum_digits` variable to 0, which will hold the sum of the digits.
- We convert the input number to a string using `str()` so that we can iterate through each digit as characters in the string.
- We use a `for` loop to iterate through each digit in the string.
- For each digit, we convert it back to an integer using `int()` and add it to the `sum_digits` variable.
- After processing all digits, the final value of `sum_digits` will be the sum of digits of the given number.

## 40. Implement a function to check if a given string is a valid palindrome considering case sensitivity.

In [41]:
def is_valid_palindrome(s):
    # Remove non-alphanumeric characters and convert the string to lowercase
    s = ''.join(c.lower() for c in s if c.isalnum())
    
    # Compare the original string with its reverse
    return s == s[::-1]

# Test the function
input_string = "A man, a plan, a canal: Panama"
result = is_valid_palindrome(input_string)
print("Is valid palindrome:", result)

Is valid palindrome: True


Explanation:
- The `is_valid_palindrome()` function takes a string `s` as input and returns `True` if it is a valid palindrome (considering case sensitivity), and `False` otherwise.
- We first remove all non-alphanumeric characters from the string and convert it to lowercase using list comprehension and `join()` method.
- Then, we compare the original string `s` with its reverse `s[::-1]`. If they are the same, the string is a valid palindrome.

## 41. Write a Python program to find the smallest missing positive integer in a list.

To find the smallest missing positive integer in a list, we can use the concept of swapping elements in place to organize the list. We traverse the list and place each positive integer `x` in its correct position at index `x-1`. After rearranging the list, we can find the smallest missing positive integer by iterating through the list and checking if the element at index `i` is equal to `i + 1`. If not, then `i + 1` is the smallest missing positive integer.

In [42]:
def first_missing_positive(nums):
    n = len(nums)
    
    # First pass to organize the list by swapping elements to their correct positions
    for i in range(n):
        while 1 <= nums[i] <= n and nums[nums[i] - 1] != nums[i]:
            nums[nums[i] - 1], nums[i] = nums[i], nums[nums[i] - 1]
    # Second pass to find the smallest missing positive integer
    for i in range(n):
        if nums[i] != i + 1:
            return i + 1
    # If all positive integers from 1 to n are present, then the smallest missing positive is n+1
    return n + 1

# TEst the function
numbers = [3, 4, -1, 1]
result = first_missing_positive(numbers)
print("Smallest missing positive integer:", result)

Smallest missing positive integer: 2


Explanation:
- The `first_missing_positive()` function takes a list of integers `nums` as input and returns the smallest missing positive integer in the list.
- In the first pass, we rearrange the list by swapping elements to their correct positions. For example, if we find a positive integer `x` at index `i`, we move it to its correct position at index `x-1` (if it's within the bounds of the list).
- In the second pass, we check if the element at index `i` is equal to `i + 1`. If not, then `i + 1` is the smallest missing positive integer.
- If all positive integers from 1 to `n` are present in the list, then the smallest missing positive is `n+1`.

## 42. Implement a function to find the longest palindrome substring in a given string.

To find the longest palindrome substring in a given string, we can use the dynamic programming approach. We can create a 2D table to store whether substrings are palindromes or not, and then use it to find the longest palindrome substring. 

In [43]:
def longest_palindrome_substring(s):
    n = len(s)
    
    # Initialize a 2D table to store whether substrings are palindromes or not
    dp = [[False] * n for _ in range(n)]
    
    # All substrings of length 1 are palindromes
    for i in range(n):
        dp [i][i] = True
        
    # Check for palindromes of length 2
    start = 0
    max_length = 1
    for i in range(n-1):
        if s[i] == s[i+1]:
            dp[i][i+1] = True
            start = i
            max_length = 2
            
    # Check for palindromes of length 3 or more
    for length in range(3, n+1):
        for i in range(n - length + 1):
            j = i + length - 1
            if s[i] == s[j] and dp[i+1][j-1]:
                dp [i][j] = True
                if length > max_length:
                    start = i
                    max_length = length
                    
    # Return the longest palinf=drome substring
    return s[start: start + max_length]

# Test the function
input_str = "babad"
result = longest_palindrome_substring(input_str)
print("Longest palindrome substring:", result)

Longest palindrome substring: bab


Explanation:
- The `longest_palindrome_substring()` function takes a string `s` as input and returns the longest palindrome substring in that string.
- It uses a dynamic programming table `dp` to store whether substrings are palindromes or not. The table is filled based on the palindromes of length 1, 2, and so on.
- The function then returns the longest palindrome substring found using the `dp` table.

## 43. Write a program to find the number of occurrences of a given element in a list.

In [44]:
def count_occurences(lst, target):
    count = 0
    for item in lst:
        if item == target:
            count += 1
    return count

# test the function
my_list = [1,2,4,6,8,3,2,1,2,4,4,6,7]
element = 2
result = count_occurences(my_list, element)
print(f"Number of occurences of {element} in the list: {result}")

Number of occurences of 2 in the list: 3


Explanation:
- The `count_occurrences()` function takes a list `lst` and a target element `target` as input and returns the number of occurrences of `target` in the list.
- It initializes a variable `count` to 0, which will keep track of the number of occurrences.
- The function then iterates through each element in the list using a for loop.
- For each element in the list, it checks if it is equal to the target element. If it is, it increments the `count` variable by 1.
- Finally, it returns the `count`, which represents the number of occurrences of the target element in the list.

## 44. Implement a function to check if a given number is a perfect number.

A perfect number is a positive integer that is equal to the sum of its proper divisors (excluding itself).

In [45]:
def is_perfect_number(num):
    if num <= 0:
        return False
    
    divisors_sum = 0
    for i in range(1, num):
        if num % i == 0:
            divisors_sum += i
            
    return divisors_sum == num

# Test the function
num = 28
result = is_perfect_number(num)
if result:
    print(f"{num} is a perfect number.")
else:
    print(f"{num} is not a perfect number.")

28 is a perfect number.


Explanation:
- The `is_perfect_number()` function takes an integer `num` as input and returns True if `num` is a perfect number, otherwise it returns False.
- It first checks if the input number is less than or equal to 0, in which case it cannot be a perfect number, so it returns False.
- The function then initializes a variable `divisors_sum` to 0, which will keep track of the sum of divisors of `num`.
- It uses a for loop to iterate from 1 to `num - 1`, as proper divisors of `num` are less than `num`.
- For each number `i` in the loop, it checks if `num` is divisible by `i` without any remainder. If it is, then `i` is a proper divisor of `num`, so it adds `i` to the `divisors_sum`.
- Finally, it returns True if the `divisors_sum` is equal to `num`, indicating that the number is a perfect number.

## 45. Write a Python program to remove all duplicates from a string.

In [46]:
def remove_suplicates(input_string):
    # Create an empty string to store the result
    result = ""
    
    # Iterate through each character in the input string
    for char in input_string:
        # If the character is not slresdy in the result string, add it
        if char not in result:
            result += char
    return result

# test the function
input_string = "Shahequa"
result_string = remove_suplicates(input_string)
print(f"Original string: {input_string}")
print(f"String with duplicates removed: {result_string}")

Original string: Shahequa
String with duplicates removed: Shaequ


Explanation:
- The `remove_duplicates()` function takes an input string as an argument and returns a new string with all duplicates removed.
- It initializes an empty string `result` to store the unique characters.
- The function iterates through each character in the input string using a for loop.
- For each character, it checks if it is not already present in the `result` string using the `not in` condition.
- If the character is not in the `result` string, it means it is a unique character, so it is added to the `result` string.
- The function continues this process for each character in the input string, and the final `result` string contains only the unique characters from the original string.
- Finally, the function returns the `result` string, which is the input string with duplicates removed.

## 46. Implement a function to find the first missing positive.

In [48]:
def first_missing_positive(nums):
    n = len(nums)
    
    # Step 1: Move each positive number to its correct position in the list
    for i in range(n):
        while 1 <= nums[i] <= n and nums[i] != nums[nums[i] - 1]:
            nums[nums[i] - 1], nums[i] = nums[i], nums[nums[i] - 1]
            
    # Step 2: Find the first number that is out of poaition (i.e., not equal to i + 1)
    for i in range(n):
        if nums[i]!= i + 1:
            return i + 1
        
    # If all numbers from 1 to n are in the list, return n + 1
    return n + 1

# test the function 
nums = [3,4,-1,1]
result = first_missing_positive(nums)
print("First missing positive integer:", result)

First missing positive integer: 2


Explanation:
- The `first_missing_positive()` function takes a list of integers `nums` as input and returns the first missing positive integer.
- The function first performs a two-step approach to rearrange the elements in the list to their correct positions:
  1. In Step 1, it moves each positive number to its correct position by swapping elements in the list.
  2. In Step 2, it finds the first number that is out of position, which is the first missing positive integer.
- After rearranging the list in the first step, the list would become [1, 3, 4, -1]. The first missing positive integer is 2, so the function correctly returns 2 as the result.

Note: The time complexity of this solution is O(n), where n is the length of the list `nums`. The function rearranges the elements and finds the first missing positive integer in linear time.